src/main/java/co/cask/gcp/bigquery/BigQuerySource.java

```diff
@@ -221,63 +159,128 @@ public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+        // Table does not exist
+        throw new IllegalArgumentException(String.format("BigQuery table %s does not exist", request.table));
       }
-      bigqueryBuilder.setProjectId(project);
-      bigquery = bigqueryBuilder.build().getService();
 
-      TableId id = TableId.of(project, request.dataset, request.table);
-      Table table = bigquery.getTable(id);
       com.google.cloud.bigquery.Schema bgSchema = table.getDefinition().getSchema();
+      List<Schema.Field> fields = getSchemaFields(bgSchema);
+      return Schema.recordOf("output", fields);
+    } catch (Exception e) {
+      throw new Exception(e.getMessage(), e);
+    }
+  }
 
-      List<Schema.Field> fields = new ArrayList<>();
-      for (Field field : bgSchema.getFields()) {
-        LegacySQLTypeName type = field.getType();
-        Schema schema;
-        StandardSQLTypeName value = type.getStandardType();
-        if (value == StandardSQLTypeName.FLOAT64) {
-          // float is a float64, so corresponding type becomes double
-          schema = Schema.of(Schema.Type.DOUBLE);
-        } else if (value == StandardSQLTypeName.BOOL) {
-          schema = Schema.of(Schema.Type.BOOLEAN);
-        } else if (value == StandardSQLTypeName.INT64) {
-          // int is a int64, so corresponding type becomes long
-          schema = Schema.of(Schema.Type.LONG);
-        } else if (value == StandardSQLTypeName.STRING) {
-          schema = Schema.of(Schema.Type.STRING);
-        } else if (value == StandardSQLTypeName.BYTES) {
-          schema = Schema.of(Schema.Type.BYTES);
-        } else if (value == StandardSQLTypeName.TIME) {
-          schema = Schema.of(Schema.LogicalType.TIME_MICROS);
-        } else if (value == StandardSQLTypeName.DATE) {
-          schema = Schema.of(Schema.LogicalType.DATE);
-        } else if (value == StandardSQLTypeName.TIMESTAMP || value == StandardSQLTypeName.DATETIME) {
-          schema = Schema.of(Schema.LogicalType.TIMESTAMP_MICROS);
-        } else {
-          // this should never happen
-          throw new UnsupportedTypeException(String.format("Big Query Sql type %s is not supported", value));
-        }
+  /**
+   * Validate output schema. This is needed because its possible that output schema is set without using
+   * {@link #getSchema(BigQuerySourceConfig)} method.
+   */
+  private void validateOutputSchema() throws IOException {
+    Table table = BigQueryUtils.getBigqueryTable(config.getServiceAccountFilePath(), config.getProject(),
+                                                 config.dataset, config.table);
+    if (table == null) {
+      // Table does not exist
+      throw new IllegalArgumentException(String.format("BigQuery table %s does not exist.", config.table));
+    }
+
+    com.google.cloud.bigquery.Schema bgSchema = table.getDefinition().getSchema();
+    if (bgSchema == null) {
+      throw new IllegalArgumentException(String.format("BigQuery table %s must not have null schema.", config.table));
```

Comment:
what does it mean when a big query table has null schema? Does it mean there is data in it but that data has no structure? What does a select * query return in these cases?

A better error message would be 'Cannot read from table [dataset].[table] because it has no schema.' It avoids the "null" language and makes it clear that the table is valid big query table, just the plugin cannot read from it.

Reply:
It can happen if the table is created without any schema and it does not have any data/schema yet.